### PR TITLE
Adding an optional 3rd argument to ContainerBuilder::register()

### PR DESCRIPTION
**Pull request by weaverryan**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | yes (I believe yes) |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | n/a |
| License | MIT |
| Doc PR | none |

This is a really easy PR to make adding service definitions via ContainerBuilder::register() a little bit more fluid. However, is this considered a BC break? Under [Using our Classes](http://symfony.com/doc/current/contributing/code/bc.html#using-our-classes) in the BC promise, you _are_ permitted to "Add an argument to an overridden method" if it has `@api` (which this does). Does that prevent us from adding a new, optional argument? Or is it ok, since a user's implementation of the overridden method will not break (because the argument is optional)?

Thanks!

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Tests/ContainerBuilderTest.php [hunk 1]
```diff
@@ -69,9 +69,11 @@ public function testDefinitions()
     public function testRegister()
     {
         $builder = new ContainerBuilder();
-        $builder->register('foo', 'Bar\FooClass');
+        $builder->register('foo', 'Bar\FooClass', array('barArg'));
```
Comment (jakzal) [inline src/Symfony/Component/DependencyInjection/Tests/ContainerBuilderTest.php]: I'd rather add a new case instead of modifying the existing one.

**Comment by stof** [comment]
> Under Using our Classes in the BC promise, you are permitted to "Add an argument to an overridden method" if it has ``@api`` (which this does). Does that prevent us from adding a new, optional argument? Or is it ok, since a user's implementation of the overridden method will not break (because the argument is optional)?

Look at the section about working on our classes. This section answers you by saying adding new arguments in an ``@api`` method is forbidden.

**Comment by weaverryan** [comment]
@stof So, the user is not allowed to add an argument to an `@api` method that they sub-class. So, this means that we *can* add an optional argument to an `@api` method, because there is no way this could break their implementation. In other words, this is *not* a BC break. Correct?

**Comment by dosten** [comment]
@weaverryan I think that @stof is talking about the "Working on our classes" section and you about the "Using our Classes", Following the BC promise when working in the Symfony code (http://symfony.com/doc/current/contributing/code/bc.html#changing-classes) you're not allowed to add an argument with a default value in a `@api` class

**Comment by weaverryan** [comment]
@dosten you're right - this change can't be made according to the BC policy.

Thanks!

**Pull request closed without merging**